From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sun, 20 Apr 2025 16:16:42 +0200
Subject: [PATCH 014/225] mmc: sdhci-sprd: Implement SD UHS tuning
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Use the existing tuning loop to support UHS tuning using
mmc_send_tuning().

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/mmc/host/sdhci-sprd.c | 52 +++++++++++++++++++++++++----------
 1 file changed, 38 insertions(+), 14 deletions(-)

diff --git a/drivers/mmc/host/sdhci-sprd.c b/drivers/mmc/host/sdhci-sprd.c
index 3584a2b31..f68953dc3 100644
--- a/drivers/mmc/host/sdhci-sprd.c
+++ b/drivers/mmc/host/sdhci-sprd.c
@@ -94,6 +94,7 @@ struct sdhci_sprd_host {
 enum sdhci_sprd_tuning_type {
 	SDHCI_SPRD_TUNING_SD_HS_CMD,
 	SDHCI_SPRD_TUNING_SD_HS_DATA,
+	SDHCI_SPRD_TUNING_SD_UHS,
 };
 
 struct sdhci_sprd_phy_cfg {
@@ -568,9 +569,15 @@ static void sdhci_sprd_hs400_enhanced_strobe(struct mmc_host *mmc,
 		     SDHCI_SPRD_REG_32_DLL_DLY);
 }
 
-static int mmc_send_tuning_cmd(struct mmc_card *card)
+static int mmc_send_tuning_cmd(struct mmc_host *mmc)
 {
-	return mmc_send_status(card, NULL);
+	struct mmc_command cmd = {};
+
+	cmd.opcode = MMC_SET_BLOCKLEN;
+	cmd.arg = 512;
+	cmd.flags = MMC_RSP_SPI_R1 | MMC_RSP_R1 | MMC_CMD_AC;
+
+	return mmc_wait_for_cmd(mmc, &cmd, 1);
 }
 
 static int mmc_send_tuning_data(struct mmc_card *card)
@@ -625,7 +632,7 @@ static int sdhci_sprd_get_best_clk_sample(struct mmc_host *mmc, u8 *value)
 }
 
 static int sdhci_sprd_tuning(struct mmc_host *mmc, struct mmc_card *card,
-			enum sdhci_sprd_tuning_type type)
+			     enum sdhci_sprd_tuning_type type, u32 opcode)
 {
 	struct sdhci_host *host = mmc_priv(mmc);
 	struct sdhci_sprd_host *sprd_host = TO_SPRD_HOST(host);
@@ -642,17 +649,22 @@ static int sdhci_sprd_tuning(struct mmc_host *mmc, struct mmc_card *card,
 
 	sdhci_reset(host, SDHCI_RESET_CMD | SDHCI_RESET_DATA);
 
-	dll_cfg = sdhci_readl(host, SDHCI_SPRD_REG_32_DLL_CFG);
-	dll_cfg &= ~SDHCI_SPRD_CPST_EN;
-	sdhci_writel(host, dll_cfg, SDHCI_SPRD_REG_32_DLL_CFG);
+	if (type != SDHCI_SPRD_TUNING_SD_UHS) {
+		dll_cfg = sdhci_readl(host, SDHCI_SPRD_REG_32_DLL_CFG);
+		dll_cfg &= ~SDHCI_SPRD_CPST_EN;
+		sdhci_writel(host, dll_cfg, SDHCI_SPRD_REG_32_DLL_CFG);
+	}
 
 	dll_dly = p[mmc->ios.timing];
 
 	for (i = 0; i <= SDHCI_SPRD_MAX_RANGE; i++) {
-		if (type == SDHCI_SPRD_TUNING_SD_HS_CMD) {
+		if (type == SDHCI_SPRD_TUNING_SD_HS_CMD ||
+		    type == SDHCI_SPRD_TUNING_SD_UHS) {
 			dll_dly &= ~SDHCI_SPRD_CMD_DLY_MASK;
 			dll_dly |= ((i << 8) & SDHCI_SPRD_CMD_DLY_MASK);
-		} else {
+		}
+		if (type == SDHCI_SPRD_TUNING_SD_HS_DATA ||
+		    type == SDHCI_SPRD_TUNING_SD_UHS) {
 			dll_dly &= ~SDHCI_SPRD_POSRD_DLY_MASK;
 			dll_dly |= ((i << 16) & SDHCI_SPRD_POSRD_DLY_MASK);
 		}
@@ -660,9 +672,11 @@ static int sdhci_sprd_tuning(struct mmc_host *mmc, struct mmc_card *card,
 		sdhci_writel(host, dll_dly, SDHCI_SPRD_REG_32_DLL_DLY);
 
 		if (type == SDHCI_SPRD_TUNING_SD_HS_CMD)
-			value[i] = !mmc_send_tuning_cmd(card);
-		else
+			value[i] = !mmc_send_tuning_cmd(mmc);
+		else if (type == SDHCI_SPRD_TUNING_SD_HS_DATA)
 			value[i] = !mmc_send_tuning_data(card);
+		else
+			value[i] = !mmc_send_tuning(mmc, opcode, NULL);
 	}
 
 	best_clk_sample = sdhci_sprd_get_best_clk_sample(mmc, value);
@@ -672,10 +686,13 @@ static int sdhci_sprd_tuning(struct mmc_host *mmc, struct mmc_card *card,
 		goto out;
 	}
 
-	if (type == SDHCI_SPRD_TUNING_SD_HS_CMD) {
+	if (type == SDHCI_SPRD_TUNING_SD_HS_CMD ||
+	    type == SDHCI_SPRD_TUNING_SD_UHS) {
 		p[mmc->ios.timing] &= ~SDHCI_SPRD_CMD_DLY_MASK;
 		p[mmc->ios.timing] |= ((best_clk_sample << 8) & SDHCI_SPRD_CMD_DLY_MASK);
-	} else {
+	}
+	if (type == SDHCI_SPRD_TUNING_SD_HS_DATA ||
+	    type == SDHCI_SPRD_TUNING_SD_UHS) {
 		p[mmc->ios.timing] &= ~(SDHCI_SPRD_POSRD_DLY_MASK);
 		p[mmc->ios.timing] |= ((best_clk_sample << 16) & SDHCI_SPRD_POSRD_DLY_MASK);
 	}
@@ -693,12 +710,17 @@ static int sdhci_sprd_tuning(struct mmc_host *mmc, struct mmc_card *card,
 
 static int sdhci_sprd_prepare_sd_hs_cmd_tuning(struct mmc_host *mmc, struct mmc_card *card)
 {
-	return sdhci_sprd_tuning(mmc, card, SDHCI_SPRD_TUNING_SD_HS_CMD);
+	return sdhci_sprd_tuning(mmc, card, SDHCI_SPRD_TUNING_SD_HS_CMD, 0);
 }
 
 static int sdhci_sprd_execute_sd_hs_data_tuning(struct mmc_host *mmc, struct mmc_card *card)
 {
-	return sdhci_sprd_tuning(mmc, card, SDHCI_SPRD_TUNING_SD_HS_DATA);
+	return sdhci_sprd_tuning(mmc, card, SDHCI_SPRD_TUNING_SD_HS_DATA, 0);
+}
+
+static int sdhci_sprd_execute_tuning(struct mmc_host *mmc, u32 opcode)
+{
+	return sdhci_sprd_tuning(mmc, NULL, SDHCI_SPRD_TUNING_SD_UHS, opcode);
 }
 
 static void sdhci_sprd_phy_param_parse(struct sdhci_sprd_host *sprd_host,
@@ -749,6 +771,8 @@ static int sdhci_sprd_probe(struct platform_device *pdev)
 		sdhci_sprd_prepare_sd_hs_cmd_tuning;
 	host->mmc_host_ops.execute_sd_hs_tuning =
 		sdhci_sprd_execute_sd_hs_data_tuning;
+	host->mmc_host_ops.execute_tuning =
+		sdhci_sprd_execute_tuning;
 
 	/*
 	 * We can not use the standard ops to change and detect the voltage
